src/yunohost/certificate.py

```diff
@@ -0,0 +1,760 @@
+
+    live_link = os.path.join(CERT_FOLDER, domain)
+
+    # If a live link (or folder) already exists
+    if os.path.exists(live_link) :
+        # If it's not a link ... expect if to be a folder
+        if not os.path.islink(live_link):
+            # Backup it and remove it
+            _backup_current_cert(domain)
+            shutil.rmtree(live_link)
+        # Else if it's a link, simply delete it
+        elif os.path.lexists(live_link):
+            os.remove(live_link)
+
+    os.symlink(new_cert_folder, live_link)
+
+    logger.info("Restarting services...")
+
+    for service in ("postfix", "dovecot", "metronome"):
+        _run_service_command("restart", service)
+
+    _run_service_command("reload", "nginx")
+
+
+def _backup_current_cert(domain):
+    logger.info("Backuping existing certificate for domain %s", domain)
+
+    cert_folder_domain = os.path.join(CERT_FOLDER, domain)
+
+    date_tag = datetime.now().strftime("%Y%m%d.%H%M%S")
+    backup_folder = "%s-backups/%s" % (cert_folder_domain, date_tag)
+
+    shutil.copytree(cert_folder_domain, backup_folder)
+
+
+def _check_domain_is_ready_for_ACME(domain):
+    public_ip = yunohost.domain.get_public_ip()
+
+    # Check if IP from DNS matches public IP
+    if not _dns_ip_match_public_ip(public_ip, domain):
+        raise MoulinetteError(errno.EINVAL, m18n.n('certmanager_domain_dns_ip_differs_from_public_ip', domain=domain))
+
+    # Check if domain seems to be accessible through HTTP ?
+    if not _domain_is_accessible_through_HTTP(public_ip, domain):
+        raise MoulinetteError(errno.EINVAL, m18n.n('certmanager_domain_http_not_working', domain=domain))
+
+
+def _dns_ip_match_public_ip(public_ip, domain):
+    try:
+        resolver = dns.resolver.Resolver()
+        resolver.nameservers = []
+        # FFDN DNS resolvers
+        # See https://www.ffdn.org/wiki/doku.php?id=formations:dns
+        resolver.nameservers.append("80.67.169.12")    # FDN
+        resolver.nameservers.append("80.67.169.40")    #
+        resolver.nameservers.append("89.234.141.66")   # ARN
+        resolver.nameservers.append("141.255.128.100") # Aquilenet
+        resolver.nameservers.append("141.255.128.101") #
+        resolver.nameservers.append("89.234.186.18")   # Grifon
+        resolver.nameservers.append("80.67.188.188")   # LDN
+        answers = resolver.query(domain, "A")
+    except (dns.resolver.NoAnswer, dns.resolver.NXDOMAIN):
+        raise MoulinetteError(errno.EINVAL, m18n.n('certmanager_error_no_A_record', domain=domain))
+
+    dns_ip = str(answers[0])
+
+    return dns_ip == public_ip
+
+
+def _domain_is_accessible_through_HTTP(ip, domain):
+    try:
+        # Check HTTP reachability
+        requests.head("http://" + ip, headers={"Host": domain})
+
+        # Check we got the ssowat header (in HTTPS)
+        r = requests.head("https://" + ip, headers={"Host": domain}, verify=False)
+        if "x-sso-wat" not in r.headers.keys():
+            return False
```

Comment:
I think we need to decide what to do with this piece of feature : 
- do some extensive and proper checks that this appears to be the good server (requires move time to dev). Using the staging/test CA of Let's Encrypt could be a proper check, though it doesn't tell what exactly is wrong
- do some checks with particular cases where it might not work. Users can still bypass them with --no-checks (in the CLI at least). That might be confusing for some users though.
- do some minimal checks that we are pretty sure are going to work for all functional setup.

I'd be in favor of the minimal checks, given that it's not a real big deal if the cert-install fails (ok, you get a message error, but that doesn't increase the rate limit, you just need to retry after you checked your config). In the initial PR, i was simply checking the http request worked (i.e. not the header presence). I doubt there will be many cases where the DNS is properly configured, and HTTP requests get answered, but it's actually not the server we're looking for.

On the other hand if would be good to have a separate feature/improvement to get a full diagnostic of features working for a given domain - not limited to DNS and HTTP but also score on mail tester for instance, working XMPP, working mucs, etc ?